src/build-remote/build-remote.cc

```diff
@@ -0,0 +1,280 @@
+#include <cstdlib>
+#include <cstring>
+#include <algorithm>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <memory>
+#include <tuple>
+#include <iomanip>
+
+#include "shared.hh"
+#include "pathlocks.hh"
+#include "globals.hh"
+#include "serve-protocol.hh"
+#include "serialise.hh"
+#include "store-api.hh"
+#include "derivations.hh"
+
+using namespace nix;
+using std::cerr;
+using std::cin;
+
+static void handle_alarm(int sig) {
+}
+
+class machine {
+    const std::vector<string> supportedFeatures;
+    const std::vector<string> mandatoryFeatures;
+
+public:
+    const string hostName;
+    const std::vector<string> systemTypes;
+    const string sshKey;
+    const unsigned long long maxJobs;
+    const unsigned long long speedFactor;
+    bool enabled;
+
+    bool allSupported(const std::vector<string> & features) const {
+        return std::all_of(features.begin(), features.end(),
+            [&](const string & feature) {
+                return std::find(supportedFeatures.begin(),
+                    supportedFeatures.end(),
+                    feature) != supportedFeatures.end() ||
+                        std::find(mandatoryFeatures.begin(),
+                            mandatoryFeatures.end(),
+                            feature) != mandatoryFeatures.end();
+            });
+    }
+
+    bool mandatoryMet(const std::vector<string> & features) const {
+        return std::all_of(mandatoryFeatures.begin(), mandatoryFeatures.end(),
+            [&](const string & feature) {
+                return std::find(features.begin(), features.end(), feature) != features.end();
+            });
+    }
+
+    machine(decltype(hostName) hostName,
+        decltype(systemTypes) systemTypes,
+        decltype(sshKey) sshKey,
+        decltype(maxJobs) maxJobs,
+        decltype(speedFactor) speedFactor,
+        decltype(supportedFeatures) supportedFeatures,
+        decltype(mandatoryFeatures) mandatoryFeatures) :
+        supportedFeatures{std::move(supportedFeatures)},
+        mandatoryFeatures{std::move(mandatoryFeatures)},
+        hostName{std::move(hostName)},
+        systemTypes{std::move(systemTypes)},
+        sshKey{std::move(sshKey)},
+        maxJobs{std::move(maxJobs)},
+        speedFactor{speedFactor == 0 ? 1 : std::move(speedFactor)},
+        enabled{true} {};
+};;
+
+static std::vector<machine> read_conf() {
+    auto conf = getEnv("NIX_REMOTE_SYSTEMS", SYSCONFDIR "/nix/machines");
+
+    auto machines = std::vector<machine>{};
+    auto confFile = std::ifstream{conf};
+    if (confFile.good()) {
+        confFile.exceptions(std::ifstream::badbit);
```

Comment:
C++ streams considered harmful, among other things due to the wacky error/exception handling. I would just do `tokenizeString(readFile(conf), "\n")`.